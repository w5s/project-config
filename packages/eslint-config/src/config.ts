export * from './config/jsonc.js';
export * from './config/imports.js';
export * from './config/node.js';
export * from './config/yml.js';
