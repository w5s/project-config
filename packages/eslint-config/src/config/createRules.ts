import { ESLintConfig } from '@w5s/dev';

export function createRules(prefix: string) {
  return ESLintConfig.renameRules({
    'no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
  }, {
    '': prefix,
  });
}
