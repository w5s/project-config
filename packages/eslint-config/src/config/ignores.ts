import fs from 'node:fs';
import nodePath from 'node:path';
import process from 'node:process';
import findUp from 'find-up';
import parseGitignore from 'parse-gitignore';
import type { Config } from '../type.js';

const getGitignore = async (cwd: string, prefix = ''): Promise<Array<string>> => {
  const gitIgnoreFile = await findUp(nodePath.join(prefix, '.gitignore'), { cwd });
  if (gitIgnoreFile != null) {
    const { patterns } = parseGitignore.parse(await fs.promises.readFile(gitIgnoreFile));
    const returnValue = patterns.map(pattern => nodePath.join(prefix, pattern));
    return returnValue;
  }
  return [];
};

export async function ignores(options: ignores.Options = {}) {
  const cwd = process.cwd();
  return [
    {
      ignores: [
        '**/node_modules',
        '**/dist',
        '**/package-lock.json',
        '**/yarn.lock',
        '**/pnpm-lock.yaml',
        '**/bun.lockb',

        '**/.docusaurus',
        '**/output',
        '**/coverage',
        '**/temp',
        '**/.temp',
        '**/tmp',
        '**/.tmp',
        '**/.history',
        '**/.vitepress/cache',
        '**/.nuxt',
        '**/.next',
        '**/.svelte-kit',
        '**/.vercel',
        '**/.changeset',
        '**/.idea',
        '**/.cache',
        '**/.output',
        '**/.vite-inspect',
        '**/.yarn',
        '**/vendor',
        '**/vendors',
        '**/*.min.*',

        '**/*.timestamp-*.mjs', // esbuild/vite temporary files

        '.modules/',
        '.go/',
        '.pnpm-store/',
        // '!.*',
        // '.venv/',
        // 'deprecated/',
        // 'test-output/',
        // 'venv/',
        // '_generated_/',

        ...await getGitignore(cwd),
        ...await getGitignore(cwd, 'android'),
        ...await getGitignore(cwd, 'ios'),
        ...(options.ignores ?? []),
      ] as Array<string>,
      name: 'w5s/ignore',
    },
  ] as const satisfies Array<Config>;
}
export namespace ignores {
  export interface Options {
    ignores?: string[];
  }
}
