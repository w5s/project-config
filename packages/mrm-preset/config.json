{
  "aliases": {
    "configure": [
      "gitignore",
      "project",
      "contributing",
      "licenses",
      "asdf",
      "release",
      "ci",
      "lang",
      "commitlint",
      "editorconfig",
      "eslint",
      "cspell",
      "vitest",
      "renovate",
      "githooks",
      "post-configure"
    ]
  }
}
