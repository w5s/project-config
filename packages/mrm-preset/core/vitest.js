const fs = require('node:fs');
const { file, packageJson } = require('mrm-core');
const pkg = require('./pkg.js');
const npm = require('./npm.js');
const project = require('./project.js');

/**
 * @param {{
 *   state: 'present'|'absent'
 * }} options
 */
function vitest({ state }) {
  const hasVitest = state === 'present';
  const hasWorkspace = pkg.hasWorkspaces(packageJson());
  const vitestScript = 'vitest run';
  const update = (/** @type {string|undefined} */ _) =>
    hasVitest ? vitestScript : _ === vitestScript ? pkg.emptyScript : _;

  // Dependencies
  npm.dependency({
    dev: true,
    name: ['vite', 'vitest'],
    state: hasVitest ? 'present' : 'absent',
  });

  /** @type {string|undefined} */
  let viteVersion;
  /** @type {string|undefined} */
  let vitestVersion;
  pkg.withPackageJson((packageFile) => {
    viteVersion = packageFile.get('devDependencies.vite');
    vitestVersion = packageFile.get('devDependencies.vitest');
    pkg.script(packageFile, {
      name: `${project.test}:root`,
      update,
      state: hasWorkspace ? 'absent' : 'present',
    });
  });

  pkg.forEachWorkspace(({ projectDir, packageFile }) => {
    const hasSrc = fs.existsSync(`${projectDir}/src`);
    const packageState = hasVitest && hasSrc ? 'present' : 'absent';
    pkg.value(packageFile, {
      path: 'devDependencies.vite',
      state: packageState,
      update: (_) => _ ?? viteVersion,
    });
    pkg.value(packageFile, {
      path: 'devDependencies.vitest',
      state: packageState,
      update: (_) => _ ?? vitestVersion,
    });
    pkg.script(packageFile, {
      name: `${project.test}:src`,
      update,
      state: packageState,
    });
    const viteConfig = file(`${projectDir}/vite.config.mts`);
    if (packageState === 'absent') {
      viteConfig.delete();
    } else if (!viteConfig.exists()) {
      viteConfig.save(
        `/// <reference types="vitest" />
import { defineConfig } from 'vite';

export default defineConfig({
  test: {},
});
`
      );
    }
  });
}

module.exports = {
  vitest,
};
