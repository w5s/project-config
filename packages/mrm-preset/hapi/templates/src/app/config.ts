/* eslint-disable no-process-env */
import * as fs from 'fs';
import * as path from 'path';
import * as Hapi from '@hapi/hapi';
import * as HapiSwagger from 'hapi-swagger';
import * as Inert from '@hapi/inert';
import * as Vision from '@hapi/vision';
import * as HapiPino from 'hapi-pino';
import * as rfs from 'rotating-file-stream';
import type * as HapiGlue from '@hapi/glue';

export const Env = {
  getPort(): string {
    return process.env.HAPI_PORT ?? '3000';
  },
  getHost(): string {
    return process.env.HAPI_HOST ?? 'localhost';
  },
  getLogFolder(): string | undefined {
    return process.env.HAPI_LOGFOLDER;
  },
  getNodeEnv(): string {
    return process.env.NODE_ENV ?? 'local';
  },
  getTLSKeyPath(): string {
    return process.env.TLS_KEY_PATH ?? '';
  },
  getTLSCertPath(): string {
    return process.env.TLS_CERT_PATH ?? '';
  },
};

const swaggerOptions: HapiSwagger.RegisterOptions = {
  info: {
    title: `TODO API`, // TODO: configure that
    version: '1.0', // TODO: configure that
  },
};

const server = {
  port: Env.getPort(),
  host: Env.getHost(),
  tls:
    Env.getPort() === '443'
      ? {
          key: Env.getTLSKeyPath().length > 0 ? fs.readFileSync(Env.getTLSKeyPath()) : '',
          cert: Env.getTLSCertPath().length > 0 ? fs.readFileSync(Env.getTLSCertPath()) : '',
        }
      : false,
  routes: {
    cors: true,
  },
};

const pad = (num: number): string => {
  return (num > 9 ? '' : '0') + String(num);
};
const generator = (time: number | Date, index?: number) => {
  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
  if (!time || !(time instanceof Date)) return path.join(Env.getLogFolder() ?? '', 'subscription_service.log');

  const year = time.getFullYear();
  const month = pad(time.getMonth() + 1);
  const day = pad(time.getDate());
  const hour = pad(time.getHours());
  const minute = pad(time.getMinutes());

  return path.join(Env.getLogFolder() ?? '', `subscription_service-${year}-${month}-${day}-${hour}-${minute}.log`);
};

const stream = rfs.createStream(generator, {
  interval: '1d',
  maxFiles: 10,
});

const plugins: Array<Hapi.ServerRegisterPluginObject<any>> = [
  {
    plugin: HapiPino,
    options: {
      prettyPrint: Env.getNodeEnv() !== 'production',
      stream: Env.getLogFolder().length > 0 ? stream : undefined,
    },
  },
  {
    plugin: Inert,
  },
  {
    plugin: Vision,
  },
  {
    plugin: HapiSwagger,
    options: swaggerOptions,
  },
];

export const serverConfig: {
  composeOptions: HapiGlue.Options;
  manifest: HapiGlue.Manifest;
} = {
  composeOptions: {
    relativeTo: __dirname,
  },
  manifest: {
    server,
    register: {
      plugins,
    },
  },
};
