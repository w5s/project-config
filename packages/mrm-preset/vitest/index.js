const { vitest } = require('../core/vitest.js');

/**
 *
 */
function task() {
  vitest({
    state: 'present',
  });
}

task.description = 'Setup Vitest';
task.parameters = {};

module.exports = task;
