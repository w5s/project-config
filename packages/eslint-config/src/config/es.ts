import globals from 'globals';
import eslintConfig from '@eslint/js';
import { Project } from '@w5s/dev';
import { type PluginOptionsBase, type Config } from '../type.js';
import type { RuleOptions } from '../typegen/jsonc.js';
import { createRules } from './createRules.js';

export async function es(
  options: es.Options,
) {
  const { rules = {} } = options;

  return [
    {
      name: 'w5s/es/setup',
      languageOptions: {
        ecmaVersion: Project.ecmaVersion(),
        globals: {
          ...globals.browser,
          ...globals[`es${Project.ecmaVersion()}`],
          ...globals.node,
          __DEV__: 'readonly',
          __PROD__: 'readonly',
          __TEST__: 'readonly',
          document: 'readonly',
          navigator: 'readonly',
          window: 'readonly',
        },
        parserOptions: {
          ecmaFeatures: {
            jsx: true,
          },
          ecmaVersion: Project.ecmaVersion(),
          sourceType: 'module',
        },
        sourceType: 'module',
      },
      linterOptions: {
        reportUnusedDisableDirectives: true,
      },
    },
    {
      name: 'w5s/es/rules',
      plugins: {},
      rules: {
        ...eslintConfig.configs.recommended.rules,
        ...createRules(''),
        ...rules,
      },
    },
  ] as [Config, Config] satisfies Array<Config>;
}

export namespace es {
  export type Rules = RuleOptions;

  export interface Options extends PluginOptionsBase<Rules> {}
}
