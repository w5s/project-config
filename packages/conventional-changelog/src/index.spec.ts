import { describe, expect, it } from 'vitest';
import config from './index.js';

describe('conventional-changelog', () => {
  it('should export data', async () => {
    await expect(config()).resolves.toMatchSnapshot();
  });
});
