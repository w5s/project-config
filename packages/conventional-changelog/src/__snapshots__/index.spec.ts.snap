// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`conventional-changelog > should export data 1`] = `
{
  "gitRawCommitOpts": {
    "format": "%B%n-hash-%n%H%n-gitTags-%n%d%n-committerDate-%n%ci%n-authorName-%n%an%n-authorEmail-%n%ae",
  },
  "parser": {
    "headerCorrespondence": [
      "type",
      "scope",
      "subject",
    ],
    "headerPattern": /\\^\\(\\?<type>\\\\S\\*\\)\\? \\(\\?:\\\\\\(\\(\\?<scope>\\.\\*\\)\\\\\\):\\? \\)\\?\\(\\?<subject>\\.\\*\\)\\$/u,
    "noteKeywords": [
      "BREAKING CHANGE",
      "BREAKING CHANGES",
    ],
    "revertCorrespondence": [
      "header",
      "hash",
    ],
    "revertPattern": /\\^\\(\\?:revert\\|revert:\\)\\\\s"\\?\\(\\[\\\\S\\\\s\\]\\+\\?\\)"\\?\\\\s\\*this reverts commit \\(\\\\w\\*\\)\\\\\\./i,
  },
  "whatBump": [Function],
  "writer": {
    "commitGroupsSort": "title",
    "commitPartial": "-{{#if scope}} **{{scope}}**:
{{~/if}} {{#if subject}}
  {{~subject}}
{{~else}}
  {{~header}}
{{~/if}}

{{~!-- commit link --}} {{#if @root.linkReferences~}}
  ([{{hash}}](
  {{~#if @root.repository}}
    {{~#if @root.host}}
      {{~@root.host}}/
    {{~/if}}
    {{~#if @root.owner}}
      {{~@root.owner}}/
    {{~/if}}
    {{~@root.repository}}
  {{~else}}
    {{~@root.repoUrl}}
  {{~/if}}/
  {{~@root.commit}}/{{hash}})) {{#if authorName}}by: **{{authorName}}**{{#if authorEmail}}<{{authorEmail}}>{{/if}}{{/if}}

{{~else}}
  {{~hash}}
{{~/if}}

{{~!-- commit references --}}
{{~#if references~}}
  , closes
  {{~#each references}} {{#if @root.linkReferences~}}
    [
    {{~#if this.owner}}
      {{~this.owner}}/
    {{~/if}}
    {{~this.repository}}#{{this.issue}}](
	{{~#if this.bugsUrl}}
      {{~this.bugsUrl}}{{this.issue}}
    {{~else}}
    {{~#if @root.repository}}
      {{~#if @root.host}}
        {{~@root.host}}/
      {{~/if}}
      {{~#if this.repository}}
        {{~#if this.owner}}
          {{~this.owner}}/
        {{~/if}}
        {{~this.repository}}
      {{~else}}
        {{~#if @root.owner}}
          {{~@root.owner}}/
        {{~/if}}
          {{~@root.repository}}
        {{~/if}}
    {{~else}}
      {{~@root.repoUrl}}
    {{~/if}}/
    {{~@root.issue}}/{{this.issue}}
	{{~/if}})
  {{~else}}
    {{~#if this.owner}}
      {{~this.owner}}/
    {{~/if}}
    {{~this.repository}}#{{this.issue}}
  {{~/if}}{{/each}}
{{~/if}}
",
    "commitsSort": [
      "scope",
      "subject",
    ],
    "footerPartial": "{{#if noteGroups}}
{{#each noteGroups}}

### {{title}}

{{#each notes}}
* {{#if commit.scope}}**{{commit.scope}}**: {{/if}}{{text}}
{{/each}}
{{/each}}

{{/if}}
",
    "groupBy": "type",
    "headerPartial": "{{#if isPatch~}}
  ##
{{~else~}}
  #
{{~/if}} {{#if @root.linkCompare~}}
  [{{version}}](
  {{~#if @root.repository~}}
    {{~#if @root.host}}
      {{~@root.host}}/
    {{~/if}}
    {{~#if @root.owner}}
      {{~@root.owner}}/
    {{~/if}}
    {{~@root.repository}}
  {{~else}}
    {{~@root.repoUrl}}
  {{~/if~}}
  /compare/{{previousTag}}...{{currentTag}})
{{~else}}
  {{~version}}
{{~/if}}
{{~#if title}} "{{title}}"
{{~/if}}
{{~#if date}} ({{date}})
{{/if}}
",
    "mainTemplate": "{{> header}}

{{#each commitGroups}}

{{#if title}}
### {{title}}

{{/if}}
{{#each commits}}
{{> commit root=@root}}{{nl}}
{{/each}}

{{/each}}
{{> footer}}
",
    "noteGroupsSort": "title",
    "transform": [Function],
  },
}
`;
